core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,399 @@
+    // verification.
+    // Setting this on either side will only do the certificate verification.
+    CertificateOnlyVerification,
+    // Setting is very DANGEROUS. Please try to avoid this in a real production environment, unless
+    // you are a super advanced user intended to re-implement the whole verification logic on your
+    // own. A secure verification might include:
+    // 1. proper verification on the peer certificate chain
+    // 2. proper checks on the identity of the peer certificate
+    InsecurelySkipAllVerification,
+  }
+
+  // Additional custom peer verification check.
+  // It will be used when checkClientTrusted/checkServerTrusted is called without
+  // {@code Socket}/{@code SSLEngine} parameter.
+  public interface PeerVerifier {
+    /**
+     * Verifies the peer certificate chain. For more information, please refer to
+     * {@code X509ExtendedTrustManager}.
+     *
+     * @param  peerCertChain  the certificate chain sent from the peer
+     * @param  authType the key exchange algorithm used, e.g. "RSA", "DHE_DSS", etc
+     */
+    void verifyPeerCertificate(X509Certificate[] peerCertChain, String authType)
+        throws CertificateException;
+  }
+
+  // Additional custom peer verification check.
+  // It will be used when checkClientTrusted/checkServerTrusted is called with the {@code Socket}
+  // parameter.
+  public interface SslSocketPeerVerifier {
+    /**
+     * Verifies the peer certificate chain. For more information, please refer to
+     * {@code X509ExtendedTrustManager}.
+     *
+     * @param  peerCertChain  the certificate chain sent from the peer
+     * @param  authType the key exchange algorithm used, e.g. "RSA", "DHE_DSS", etc
+     * @param  socket the socket used for this connection. This parameter can be null, which
+     *         indicates that implementations need not check the ssl parameters
+     */
+    void verifyPeerCertificate(X509Certificate[] peerCertChain,  String authType, Socket socket)
+        throws CertificateException;
+  }
+
+  // Additional custom peer verification check.
+  // It will be used when checkClientTrusted/checkServerTrusted is called with the {@code SSLEngine}
+  // parameter.
+  public interface SslEnginePeerVerifier {
+    /**
+     * Verifies the peer certificate chain. For more information, please refer to
+     * {@code X509ExtendedTrustManager}.
+     *
+     * @param  peerCertChain  the certificate chain sent from the peer
+     * @param  authType the key exchange algorithm used, e.g. "RSA", "DHE_DSS", etc
+     * @param  engine the engine used for this connection. This parameter can be null, which
+     *         indicates that implementations need not check the ssl parameters
+     */
+    void verifyPeerCertificate(X509Certificate[] peerCertChain, String authType, SSLEngine engine)
+        throws CertificateException;
+  }
+
+  public static final class Builder {
```

Comment:
Have a private/protected constructor? Right now users can do `newBuilder()` and `new Builder()` and get different results.

Reply:
Yeah, that's a great suggestion!